File: art/ImageProcessorModule.java
package com.coloringgame;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.Paint;
import android.net.Uri;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.util.Random;

public class ImageProcessorModule extends ReactContextBaseJavaModule {
    private static final String MODULE_NAME = "ImageProcessor";
    private final ReactApplicationContext reactContext;

    public ImageProcessorModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public String getName() {
        return MODULE_NAME;
    }

    @ReactMethod
    public void processImageToOutline(String imageUri, int targetWidth, int targetHeight, Promise promise) {
        try {
            // Load image
            Uri uri = Uri.parse(imageUri);
            InputStream inputStream = reactContext.getContentResolver().openInputStream(uri);
            Bitmap originalBitmap = BitmapFactory.decodeStream(inputStream);
            inputStream.close();

            if (originalBitmap == null) {
                promise.reject("IMAGE_LOAD_ERROR", "Failed to load image");
                return;
            }

            // Resize bitmap
            Bitmap resizedBitmap = Bitmap.createScaledBitmap(originalBitmap, targetWidth, targetHeight, true);
            originalBitmap.recycle();

            // Convert to grayscale
            Bitmap grayscaleBitmap = convertToGrayscale(resizedBitmap);

            // Apply edge detection
            Bitmap edgeBitmap = detectEdges(grayscaleBitmap);

            // Find regions
            WritableArray regions = findRegions(edgeBitmap);

            // Save processed image
            String processedPath = saveImageToTemp(edgeBitmap);

            // Create result
            WritableMap result = Arguments.createMap();
            result.putString("outlineUri", processedPath);
            result.putInt("width", targetWidth);
            result.putInt("height", targetHeight);
            result.putArray("regions", regions);

            // Cleanup
            resizedBitmap.recycle();
            grayscaleBitmap.recycle();
            edgeBitmap.recycle();

            promise.resolve(result);

        } catch (Exception e) {
            promise.reject("PROCESSING_ERROR", e.getMessage(), e);
        }
    }

    private Bitmap convertToGrayscale(Bitmap bitmap) {
        Bitmap grayscaleBitmap = Bitmap.createBitmap(bitmap.getWidth(), bitmap.getHeight(), Bitmap.Config.ARGB_8888);
        
        Canvas canvas = new Canvas(grayscaleBitmap);
        Paint paint = new Paint();
        
        ColorMatrix colorMatrix = new ColorMatrix();
        colorMatrix.setSaturation(0);
        
        ColorMatrixColorFilter filter = new ColorMatrixColorFilter(colorMatrix);
        paint.setColorFilter(filter);
        
        canvas.drawBitmap(bitmap, 0, 0, paint);
        
        return grayscaleBitmap;
    }

    private Bitmap detectEdges(Bitmap bitmap) {
        int width = bitmap.getWidth();
        int height = bitmap.getHeight();
        
        Bitmap edgeBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
        
        int[] pixels = new int[width * height];
        bitmap.getPixels(pixels, 0, width, 0, 0, width, height);
        
        int[] edgePixels = new int[width * height];
        
        // Sobel operator
        int[][] sobelX = {{-1, 0, 1}, {-2, 0, 2}, {-1, 0, 1}};
        int[][] sobelY = {{-1, -2, -1}, {0, 0, 0}, {1, 2, 1}};
        
        for (int y = 1; y < height - 1; y++) {
            for (int x = 1; x < width - 1; x++) {
                int gx = 0;
                int gy = 0;
                
                for (int ky = -1; ky <= 1; ky++) {
                    for (int kx = -1; kx <= 1; kx++) {
                        int pixel = pixels[(y + ky) * width + (x + kx)];
                        int gray = (pixel >> 16) & 0xFF; // Red channel (all channels same in grayscale)
                        
                        gx += gray * sobelX[ky + 1][kx + 1];
                        gy += gray * sobelY[ky + 1][kx + 1];
                    }
                }
                
                int magnitude = (int) Math.sqrt(gx * gx + gy * gy);
                magnitude = Math.min(255, magnitude);
                
                int edgeColor = (magnitude > 128) ? 0xFF000000 : 0xFFFFFFFF; // Black or white
                edgePixels[y * width + x] = edgeColor;
            }
        }
        
        edgeBitmap.setPixels(edgePixels, 0, width, 0, 0, width, height);
        
        return edgeBitmap;
    }

    private WritableArray findRegions(Bitmap edgeBitmap) {
        WritableArray regions = Arguments.createArray();
        
        int width = edgeBitmap.getWidth();
        int height = edgeBitmap.getHeight();
        
        // Simple grid-based region detection
        int gridSize = 5;
        int cellWidth = width / gridSize;
        int cellHeight = height / gridSize;
        
        Random random = new Random();
        
        for (int row = 0; row < gridSize; row++) {
            for (int col = 0; col < gridSize; col++) {
                int x = col * cellWidth + cellWidth / 2;
                int y = row * cellHeight + cellHeight / 2;
                
                // Add random offset
                x += random.nextInt(cellWidth / 2) - cellWidth / 4;
                y += random.nextInt(cellHeight / 2) - cellHeight / 4;
                
                WritableMap center = Arguments.createMap();
                center.putInt("x", x);
                center.putInt("y", y);
                
                WritableMap region = Arguments.createMap();
                region.putInt("id", row * gridSize + col);
                region.putMap("center", center);
                region.putInt("colorId", (row * gridSize + col) % 10 + 1);
                
                regions.pushMap(region);
            }
        }
        
        return regions;
    }

    private String saveImageToTemp(Bitmap bitmap) throws Exception {
        File tempDir = reactContext.getCacheDir();
        String filename = "processed_" + System.currentTimeMillis() + ".png";
        File file = new File(tempDir, filename);
        
        FileOutputStream outputStream = new FileOutputStream(file);
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream);
        outputStream.flush();
        outputStream.close();
        
        return "file://" + file.getAbsolutePath();
    }
}

File: art/ImageProcessorModule.m
//
//  ImageProcessorModule.m
//  ColorByNumbersGame
//
//  Native module implementation for image processing
//

#import "ImageProcessorModule.h"
#import <React/RCTLog.h>
#import <UIKit/UIKit.h>

// If using OpenCV, uncomment:
// #ifdef __cplusplus
// #import <opencv2/opencv.hpp>
// #import <opencv2/imgcodecs/ios.h>
// #endif

@implementation ImageProcessorModule

RCT_EXPORT_MODULE(ImageProcessor);

// Process image to outline
RCT_EXPORT_METHOD(processImageToOutline:(NSString *)imageUri
                  targetWidth:(NSInteger)width
                  targetHeight:(NSInteger)height
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        @try {
            // Load image
            NSURL *url = [NSURL URLWithString:imageUri];
            NSData *imageData = [NSData dataWithContentsOfURL:url];
            UIImage *image = [UIImage imageWithData:imageData];
            
            if (!image) {
                reject(@"image_load_error", @"Failed to load image", nil);
                return;
            }
            
            // Resize image
            UIImage *resizedImage = [self resizeImage:image toWidth:width height:height];
            
            // Convert to grayscale
            UIImage *grayscaleImage = [self convertToGrayscale:resizedImage];
            
            // Apply edge detection (simplified Sobel)
            UIImage *edgeImage = [self detectEdges:grayscaleImage];
            
            // Find regions
            NSArray *regions = [self findRegions:edgeImage];
            
            // Save processed image
            NSString *processedPath = [self saveImageToTemp:edgeImage];
            
            // Return results
            NSDictionary *result = @{
                @"outlineUri": processedPath,
                @"width": @(width),
                @"height": @(height),
                @"regions": regions
            };
            
            dispatch_async(dispatch_get_main_queue(), ^{
                resolve(result);
            });
            
        } @catch (NSException *exception) {
            reject(@"processing_error", exception.reason, nil);
        }
    });
}

// Resize image
- (UIImage *)resizeImage:(UIImage *)image toWidth:(NSInteger)width height:(NSInteger)height {
    CGSize newSize = CGSizeMake(width, height);
    UIGraphicsBeginImageContextWithOptions(newSize, NO, 1.0);
    [image drawInRect:CGRectMake(0, 0, newSize.width, newSize.height)];
    UIImage *resizedImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return resizedImage;
}

// Convert to grayscale
- (UIImage *)convertToGrayscale:(UIImage *)image {
    CGSize size = image.size;
    CGRect rect = CGRectMake(0, 0, size.width, size.height);
    
    // Create a grayscale context
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceGray();
    CGContextRef context = CGBitmapContextCreate(nil, size.width, size.height, 8, 0, colorSpace, kCGImageAlphaNone);
    
    // Draw the image
    CGContextDrawImage(context, rect, image.CGImage);
    
    // Get the grayscale image
    CGImageRef imageRef = CGBitmapContextCreateImage(context);
    UIImage *grayscaleImage = [UIImage imageWithCGImage:imageRef];
    
    // Cleanup
    CGColorSpaceRelease(colorSpace);
    CGContextRelease(context);
    CGImageRelease(imageRef);
    
    return grayscaleImage;
}

// Simple edge detection (Sobel-like)
- (UIImage *)detectEdges:(UIImage *)image {
    CIContext *context = [CIContext contextWithOptions:nil];
    CIImage *ciImage = [CIImage imageWithCGImage:image.CGImage];
    
    // Apply edge detection filter
    CIFilter *edgeFilter = [CIFilter filterWithName:@"CIEdges"];
    [edgeFilter setValue:ciImage forKey:kCIInputImageKey];
    [edgeFilter setValue:@(2.0) forKey:kCIInputIntensityKey];
    
    CIImage *outputImage = [edgeFilter outputImage];
    
    // Convert back to UIImage
    CGImageRef cgImage = [context createCGImage:outputImage fromRect:outputImage.extent];
    UIImage *edgeImage = [UIImage imageWithCGImage:cgImage];
    CGImageRelease(cgImage);
    
    return edgeImage;
}

// Find regions using connected components
- (NSArray *)findRegions:(UIImage *)edgeImage {
    NSMutableArray *regions = [NSMutableArray array];
    
    CGImageRef imageRef = edgeImage.CGImage;
    NSInteger width = CGImageGetWidth(imageRef);
    NSInteger height = CGImageGetHeight(imageRef);
    
    // Get pixel data
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceGray();
    unsigned char *pixels = (unsigned char *)calloc(width * height, sizeof(unsigned char));
    
    CGContextRef contextRef = CGBitmapContextCreate(pixels, width, height, 8, width, colorSpace, kCGImageAlphaNone);
    CGContextDrawImage(contextRef, CGRectMake(0, 0, width, height), imageRef);
    
    // Simple region detection (grid-based for performance)
    NSInteger gridSize = 5;
    NSInteger cellWidth = width / gridSize;
    NSInteger cellHeight = height / gridSize;
    
    for (NSInteger row = 0; row < gridSize; row++) {
        for (NSInteger col = 0; col < gridSize; col++) {
            NSInteger x = col * cellWidth + cellWidth / 2;
            NSInteger y = row * cellHeight + cellHeight / 2;
            
            // Add some random offset
            x += (arc4random() % (cellWidth / 2)) - cellWidth / 4;
            y += (arc4random() % (cellHeight / 2)) - cellHeight / 4;
            
            NSDictionary *region = @{
                @"id": @(row * gridSize + col),
                @"center": @{
                    @"x": @(x),
                    @"y": @(y)
                },
                @"colorId": @((row * gridSize + col) % 10 + 1)
            };
            
            [regions addObject:region];
        }
    }
    
    // Cleanup
    free(pixels);
    CGColorSpaceRelease(colorSpace);
    CGContextRelease(contextRef);
    
    return regions;
}

// Save image to temporary directory
- (NSString *)saveImageToTemp:(UIImage *)image {
    NSString *tempDir = NSTemporaryDirectory();
    NSString *filename = [NSString stringWithFormat:@"processed_%@.png", [[NSUUID UUID] UUIDString]];
    NSString *filePath = [tempDir stringByAppendingPathComponent:filename];
    
    NSData *imageData = UIImagePNGRepresentation(image);
    [imageData writeToFile:filePath atomically:YES];
    
    return [NSString stringWithFormat:@"file://%@", filePath];
}

@end
